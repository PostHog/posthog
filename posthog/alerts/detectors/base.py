"""Base classes and types for alert detectors."""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from enum import StrEnum
from typing import Any, Optional


class DetectorType(StrEnum):
    """Available detector types."""

    THRESHOLD = "threshold"
    ZSCORE = "zscore"
    MAD = "mad"  # Median Absolute Deviation


class DetectionDirection(StrEnum):
    """Direction for anomaly detection."""

    UP = "up"  # Alert on values above normal
    DOWN = "down"  # Alert on values below normal
    BOTH = "both"  # Alert on values above or below normal


class ValueType(StrEnum):
    """Whether to run detection on raw values or deltas."""

    RAW = "raw"  # Run detection on raw metric values
    DELTA = "delta"  # Run detection on period-over-period changes


@dataclass
class DetectionResult:
    """Result of running a detector on data."""

    value: Optional[float]  # The raw metric value
    detector_score: Optional[float]  # The detector-specific score (e.g., z-score, mad score)
    is_breach: bool  # Whether this constitutes a breach
    breach_messages: list[str]  # Human-readable breach descriptions
    metadata: dict[str, Any]  # Additional detector-specific info


class BaseDetector(ABC):
    """Base class for all alert detectors."""

    def __init__(self, config: dict[str, Any]):
        """Initialize detector with configuration."""
        self.config = config
        self.validate_config()

    @abstractmethod
    def detect(
        self, values: list[float], series_name: str = "Series", value_type: ValueType = ValueType.RAW
    ) -> DetectionResult:
        """
        Run detection on a list of values.

        Args:
            values: Time series values (most recent last)
            series_name: Name of the series for breach messages
            value_type: Whether to use raw values or deltas

        Returns:
            DetectionResult with breach status and details
        """
        pass

    @abstractmethod
    def validate_config(self) -> None:
        """Validate the detector configuration. Raise ValueError if invalid."""
        pass

    @property
    @abstractmethod
    def detector_type(self) -> DetectorType:
        """Return the type of this detector."""
        pass

    def _compute_deltas(self, values: list[float]) -> list[float]:
        """Compute period-over-period deltas from values."""
        if len(values) < 2:
            return []
        return [values[i] - values[i - 1] for i in range(1, len(values))]
