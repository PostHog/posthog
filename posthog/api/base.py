from rest_framework.pagination import CursorPagination as RestCursorPagination


class CursorPagination(RestCursorPagination):
    ordering = "-created_at"
    page_size = 100
