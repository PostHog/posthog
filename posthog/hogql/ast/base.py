import re
from typing import Optional, Literal

from pydantic import BaseModel, Extra

from posthog.hogql.constants import ConstantDataType
from posthog.hogql.errors import NotImplementedException

# NOTE: when you add new AST fields or nodes, add them to the Visitor classes in visitor.py as well!

camel_case_pattern = re.compile(r"(?<!^)(?=[A-Z])")


class AST(BaseModel):
    class Config:
        extra = Extra.forbid

    def accept(self, visitor):
        camel_case_name = camel_case_pattern.sub("_", self.__class__.__name__).lower()
        method_name = f"visit_{camel_case_name}"
        if hasattr(visitor, method_name):
            visit = getattr(visitor, method_name)
            return visit(self)
        if hasattr(visitor, "visit_unknown"):
            return visitor.visit_unknown(self)
        raise NotImplementedException(f"Visitor has no method {method_name}")


class Type(AST):
    def get_child(self, name: str) -> "Type":
        raise NotImplementedException("Type.get_child not overridden")

    def has_child(self, name: str) -> bool:
        return self.get_child(name) is not None

    def resolve_constant_type(self) -> Optional["ConstantType"]:
        from posthog.hogql.ast import UnknownType

        return UnknownType()


class ConstantType(Type):
    data_type: ConstantDataType

    def resolve_constant_type(self) -> "ConstantType":
        return self


class Expr(AST):
    type: Optional[Type]


class Macro(Expr):
    name: str
    expr: Expr
    # Whether the macro is an inlined column "SELECT 1 AS a" or a subquery "SELECT a AS (SELECT 1)"
    macro_format: Literal["column", "subquery"]
