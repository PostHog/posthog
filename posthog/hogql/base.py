import re
from dataclasses import dataclass, field

from typing import Literal, Optional

from posthog.hogql.constants import ConstantDataType
from posthog.hogql.errors import NotImplementedException


# Given a string like "CorrectHorseBS", match the "H" and "B", so that we can convert this to "correct_horse_bs"
camel_case_pattern = re.compile(r"(?<!^)(?<![A-Z])(?=[A-Z])")


@dataclass(kw_only=True)
class AST:
    start: Optional[int] = field(default=None)
    end: Optional[int] = field(default=None)

    def accept(self, visitor):
        camel_case_name = camel_case_pattern.sub("_", self.__class__.__name__).lower()
        if "hog_qlx" in camel_case_name:
            camel_case_name = camel_case_name.replace("hog_qlx", "hogqlx_")
        method_name = f"visit_{camel_case_name}"
        if hasattr(visitor, method_name):
            visit = getattr(visitor, method_name)
            return visit(self)
        if hasattr(visitor, "visit_unknown"):
            return visitor.visit_unknown(self)
        raise NotImplementedException(f"Visitor has no method {method_name}")


@dataclass(kw_only=True)
class Type(AST):
    def get_child(self, name: str) -> "Type":
        raise NotImplementedException("Type.get_child not overridden")

    def has_child(self, name: str) -> bool:
        return self.get_child(name) is not None

    def resolve_constant_type(self) -> Optional["ConstantType"]:
        return UnknownType()


@dataclass(kw_only=True)
class Expr(AST):
    type: Optional[Type] = field(default=None)


@dataclass(kw_only=True)
class CTE(Expr):
    """A common table expression."""

    name: str
    expr: Expr
    # Whether the CTE is an inlined column "WITH 1 AS a" or a subquery "WITH a AS (SELECT 1)"
    cte_type: Literal["column", "subquery"]


@dataclass(kw_only=True)
class ConstantType(Type):
    data_type: ConstantDataType

    def resolve_constant_type(self) -> "ConstantType":
        return self

    def print_type(self) -> str:
        raise NotImplementedException("ConstantType.print_type not implemented")


@dataclass(kw_only=True)
class UnknownType(ConstantType):
    data_type: ConstantDataType = field(default="unknown", init=False)

    def print_type(self) -> str:
        return "Unknown"
