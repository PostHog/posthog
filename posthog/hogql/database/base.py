from typing import Any, Callable, Dict, List

from pydantic import BaseModel, Extra

from posthog.hogql.errors import HogQLException, NotImplementedException
from posthog.hogql.database.fields import DatabaseField


class Table(BaseModel):
    class Config:
        extra = Extra.forbid

    def has_field(self, name: str) -> bool:
        return hasattr(self, name)

    def get_field(self, name: str) -> DatabaseField:
        if self.has_field(name):
            return getattr(self, name)
        raise HogQLException(f'Field "{name}" not found on table {self.__class__.__name__}')

    def clickhouse_table(self):
        raise NotImplementedException("Table.clickhouse_table not overridden")

    def hogql_table(self):
        raise NotImplementedException("Table.hogql_table not overridden")

    def avoid_asterisk_fields(self) -> List[str]:
        return []

    def get_asterisk(self) -> Dict[str, DatabaseField]:
        asterisk: Dict[str, DatabaseField] = {}
        fields_to_avoid = self.avoid_asterisk_fields() + ["team_id"]
        for key, field in self.__fields__.items():
            if key in fields_to_avoid:
                continue
            database_field = field.default
            if isinstance(database_field, DatabaseField):
                asterisk[key] = database_field
            elif (
                isinstance(database_field, Table)
                or isinstance(database_field, LazyJoin)
                or isinstance(database_field, FieldTraverser)
            ):
                pass  # ignore virtual tables for now
            else:
                raise HogQLException(f"Unknown field type {type(database_field).__name__} for asterisk")
        return asterisk


class LazyJoin(BaseModel):
    class Config:
        extra = Extra.forbid

    join_function: Callable[[str, str, Dict[str, Any]], Any]
    join_table: Table
    from_field: str


class LazyTable(Table):
    class Config:
        extra = Extra.forbid

    def lazy_select(self, requested_fields: Dict[str, Any]) -> Any:
        raise NotImplementedException("LazyTable.lazy_select not overridden")


class VirtualTable(Table):
    class Config:
        extra = Extra.forbid


class FieldTraverser(BaseModel):
    class Config:
        extra = Extra.forbid

    chain: List[str]
