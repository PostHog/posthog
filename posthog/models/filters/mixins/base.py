from typing import Dict, Literal

BreakdownType = Literal["event", "person", "cohort", "group"]
IntervalType = Literal["hour", "day", "week", "month"]
FunnelWindowIntervalType = Literal["minute", "hour", "day", "week", "month"]


class BaseParamMixin:
    _data: Dict
