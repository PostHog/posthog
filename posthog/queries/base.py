from typing import Any, Callable, Dict, List, Union, cast

from dateutil.relativedelta import relativedelta
from django.db.models import Exists, OuterRef, Q
from rest_framework.exceptions import ValidationError

from posthog.models.cohort import Cohort
from posthog.models.filters.filter import Filter
from posthog.models.person import Person
from posthog.models.property import Property
from posthog.models.team import Team
from posthog.utils import get_compare_period_dates


def determine_compared_filter(filter) -> Filter:
    if not filter.date_to or not filter.date_from:
        raise ValidationError("You need date_from and date_to to compare")
    date_from, date_to = get_compare_period_dates(filter.date_from, filter.date_to)

    date_from += relativedelta(days=1)
    return filter.with_data({"date_from": date_from.date().isoformat(), "date_to": date_to.date().isoformat()})


def convert_to_comparison(trend_entity: List[Dict[str, Any]], filter, label: str) -> List[Dict[str, Any]]:
    for entity in trend_entity:
        labels = [
            "{} {}".format(filter.interval if filter.interval is not None else "day", i)
            for i in range(len(entity["labels"]))
        ]
        entity.update(
            {
                "labels": labels,
                "days": entity["days"],
                "label": entity["label"],
                "compare_label": label,
                "compare": True,
            }
        )
    return trend_entity


"""
    handle_compare takes an Entity, Filter and a callable.
    It'll automatically create a new entity with the 'current' and 'previous' labels and automatically pick the right date_from and date_to filters .
    It will then call func(entity, filter, team_id).
"""


def handle_compare(filter, func: Callable, team: Team, **kwargs) -> List:
    entities_list = []
    trend_entity = func(filter=filter, team=team, **kwargs)
    if filter.compare:
        trend_entity = convert_to_comparison(trend_entity, filter, "current")
        entities_list.extend(trend_entity)

        compared_filter = determine_compared_filter(filter)
        compared_trend_entity = func(filter=compared_filter, team=team, **kwargs)

        compared_trend_entity = convert_to_comparison(compared_trend_entity, compared_filter, "previous",)
        entities_list.extend(compared_trend_entity)
    else:
        entities_list.extend(trend_entity)
    return entities_list


TIME_IN_SECONDS: Dict[str, Any] = {
    "hour": 3600,
    "day": 3600 * 24,
    "week": 3600 * 24 * 7,
    "month": 3600 * 24 * 30,  # TODO: Let's get rid of this lie! Months are not all 30 days long
}


def properties_to_Q(properties: List[Property], team_id: int, is_direct_query: bool = False) -> Q:
    """
    Converts a filter to Q, for use in Django ORM .filter()
    If you're filtering a Person/Group QuerySet, use is_direct_query to avoid doing an unnecessary nested loop
    """
    filters = Q()

    if len(properties) == 0:
        return filters

    if is_direct_query:
        for property in properties:
            filters &= property.property_to_Q()
        return filters

    person_properties = [prop for prop in properties if prop.type == "person"]
    if len(person_properties) > 0:
        person_Q = Q()
        for property in person_properties:
            person_Q &= property.property_to_Q()
        filters &= Q(Exists(Person.objects.filter(person_Q, id=OuterRef("person_id"),).only("pk")))

    event_properties = [prop for prop in properties if prop.type == "event"]
    if len(event_properties) > 0:
        raise ValueError("Event properties are no longer supported in properties_to_Q")

    # importing from .event and .cohort below to avoid importing from partially initialized modules

    element_properties = [prop for prop in properties if prop.type == "element"]
    if len(element_properties) > 0:
        raise ValueError("Element properties are no longer supported in properties_to_Q")

    cohort_properties = [prop for prop in properties if prop.type == "cohort"]
    if len(cohort_properties) > 0:
        from posthog.models.cohort import CohortPeople

        for item in cohort_properties:
            if item.key == "id":
                cohort_id = int(cast(Union[str, int], item.value))
                cohort = Cohort.objects.get(pk=cohort_id)
                filters &= Q(
                    Exists(
                        CohortPeople.objects.filter(
                            cohort_id=cohort.pk, person_id=OuterRef("person_id"), version=cohort.version
                        ).only("id")
                    )
                )

    if len([prop for prop in properties if prop.type == "group"]):
        raise ValueError("Group properties are not supported for indirect filtering via postgres")

    return filters
