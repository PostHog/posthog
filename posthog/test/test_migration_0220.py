import pytest

from posthog.test.base import TestMigrations

pytestmark = pytest.mark.skip("old migrations slow overall test run down")


class TagsTestCase(TestMigrations):

    migrate_from = "0219_migrate_tags_v2"  # type: ignore
    migrate_to = "0220_backfill_primary_dashboards"  # type: ignore
    assert_snapshots = True

    def setUpBeforeMigration(self, apps):
        Organization = apps.get_model("posthog", "Organization")
        Dashboard = apps.get_model("posthog", "Dashboard")
        Team = apps.get_model("posthog", "Team")

        org = Organization.objects.create(name="o1")

        # CASE 1:
        # Team with an existing primary dashboard, and non primary dashboards
        # Expect: t1.primary_dashboard = d2 (even though d1 is created first)
        team_1 = Team.objects.create(name="t1", organization=org)
        Dashboard.objects.create(name="d1", team=team_1)
        dashboard_2 = Dashboard.objects.create(name="d2", team=team_1)
        team_1.primary_dashboard = dashboard_2
        team_1.save()

        # CASE 2:
        # Team with no dashboards
        # Expect: t2.primary_dashboard = None
        Team.objects.create(name="t2", organization=org)

        # CASE 3:
        # Team with multiple pinned and non pinned dashboards, and no primary set
        # Expect: t3.primary_dashboard = d4 (even though d3 is created first)
        team_3 = Team.objects.create(name="t3", organization=org)
        Dashboard.objects.create(name="d3", team=team_3, pinned=False)
        Dashboard.objects.create(name="d4", team=team_3, pinned=True)
        Dashboard.objects.create(name="d5", team=team_3, pinned=True)

        # CASE 4:
        # Team with multiple non-pinned dashboards, and no primary set
        # Expect: t4.primary_dashboard = d6
        team_4 = Team.objects.create(name="t4", organization=org)
        Dashboard.objects.create(name="d6", team=team_4)
        Dashboard.objects.create(name="d7", team=team_4)

        # BATCH CASE
        teams = Team.objects.bulk_create(
            [Team(name=f"batch_team-{team_number+10}", organization=org) for team_number in range(501)]
        )
        Dashboard.objects.bulk_create(
            [
                Dashboard(name=f"batch_dashboard-{dashboard_number+10}", team=team,)
                for dashboard_number, team in enumerate(teams)
            ]
        )

    def test_backfill_primary_dashboard(self):
        Dashboard = self.apps.get_model("posthog", "Dashboard")  # type: ignore
        Team = self.apps.get_model("posthog", "Team")  # type: ignore

        # CASE 1:
        self.assertEqual(Team.objects.get(name="t1").primary_dashboard.id, Dashboard.objects.get(name="d2").id)

        # CASE 2:
        self.assertEqual(Team.objects.get(name="t2").primary_dashboard, None)

        # CASE 3:
        self.assertEqual(Team.objects.get(name="t3").primary_dashboard.id, Dashboard.objects.get(name="d4").id)

        # CASE 4:
        self.assertEqual(Team.objects.get(name="t4").primary_dashboard.id, Dashboard.objects.get(name="d6").id)

        # BATCH CASE
        teams = Team.objects.filter(name__startswith="batch_team-")
        self.assertEqual(teams.count(), 501)
        for team in teams:
            team_number = team.name.split("-")[1]
            self.assertEqual(team.primary_dashboard.name.split("-")[1], team_number)

    def tearDown(self):
        Team = self.apps.get_model("posthog", "Team")  # type: ignore
        Dashboard = self.apps.get_model("posthog", "Dashboard")  # type: ignore
        Dashboard.objects.all().delete()
        Team.objects.all().delete()
        super().tearDown()
