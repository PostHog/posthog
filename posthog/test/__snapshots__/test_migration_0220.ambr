# name: TagsTestCase.test_backfill_primary_dashboard
  '
  
  SELECT c.relname,
         CASE
             WHEN c.relispartition THEN 'p'
             WHEN c.relkind IN ('m',
                                'v') THEN 'v'
             ELSE 't'
         END
  FROM pg_catalog.pg_class c
  LEFT JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
  WHERE c.relkind IN ('f',
                      'm',
                      'p',
                      'r',
                      'v')
    AND n.nspname NOT IN ('pg_catalog',
                          'pg_toast')
    AND pg_catalog.pg_table_is_visible(c.oid)
  '
---
# name: TagsTestCase.test_backfill_primary_dashboard.1
  '
  
  SELECT posthog_team.id,
         COALESCE(MIN(CASE
                          WHEN posthog_dashboard.pinned THEN posthog_dashboard.id
                          ELSE NULL
                      END), MIN(CASE
                                    WHEN NOT posthog_dashboard.pinned THEN posthog_dashboard.id
                                    ELSE NULL
                                END)) AS primary_dashboard_id
  FROM posthog_team
  INNER JOIN posthog_dashboard ON posthog_dashboard.team_id = posthog_team.id
  WHERE posthog_team.primary_dashboard_id IS NULL
  GROUP BY posthog_team.id
  '
---
# name: TagsTestCase.test_backfill_primary_dashboard.2
  '
  SELECT "posthog_team"."id",
         "posthog_team"."primary_dashboard_id"
  FROM "posthog_team"
  WHERE "posthog_team"."id" IN (1,
                                2,
                                3,
                                4,
                                5 /* ... */)
  '
---
# name: TagsTestCase.test_backfill_primary_dashboard.3
  '
  SELECT "posthog_team"."id",
         "posthog_team"."primary_dashboard_id"
  FROM "posthog_team"
  WHERE "posthog_team"."id" IN (1,
                                2,
                                3,
                                4,
                                5 /* ... */)
  '
---
# name: TagsTestCase.test_backfill_primary_dashboard.4
  '
  
  SELECT c.relname,
         CASE
             WHEN c.relispartition THEN 'p'
             WHEN c.relkind IN ('m',
                                'v') THEN 'v'
             ELSE 't'
         END
  FROM pg_catalog.pg_class c
  LEFT JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
  WHERE c.relkind IN ('f',
                      'm',
                      'p',
                      'r',
                      'v')
    AND n.nspname NOT IN ('pg_catalog',
                          'pg_toast')
    AND pg_catalog.pg_table_is_visible(c.oid)
  '
---
# name: TagsTestCase.test_backfill_primary_dashboard.5
  '
  
  SELECT c.relname,
         CASE
             WHEN c.relispartition THEN 'p'
             WHEN c.relkind IN ('m',
                                'v') THEN 'v'
             ELSE 't'
         END
  FROM pg_catalog.pg_class c
  LEFT JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
  WHERE c.relkind IN ('f',
                      'm',
                      'p',
                      'r',
                      'v')
    AND n.nspname NOT IN ('pg_catalog',
                          'pg_toast')
    AND pg_catalog.pg_table_is_visible(c.oid)
  '
---
# name: TagsTestCase.test_backfill_primary_dashboard.6
  '
  SELECT "django_migrations"."id",
         "django_migrations"."app",
         "django_migrations"."name",
         "django_migrations"."applied"
  FROM "django_migrations"
  '
---
